File: apps/player/apps/native/Sources/Services/RemotePlayersAPI.swift
import Foundation

struct RemotePlayerCommand: Sendable {
    let action: String
    let itemID: String
    let position: Double?

    init(_ raw: JSONValue) throws {
        let value = try raw.object(allowing: ["command", "itemId", "position"])
        action = try value.text("command", max: 16, required: true)
        itemID = try Input.id(value.text("itemId", max: 128, required: true))
        guard ["play", "pause", "backward", "forward", "previous", "next", "seek"].contains(action) else { throw ClientError.invalidResponse }
        if action == "seek" {
            guard value["position"] != nil else { throw ClientError.invalidResponse }
            position = try value.number("position", max: 1_000_000_000)
        } else {
            guard value["position"] == nil else { throw ClientError.invalidResponse }
            position = nil
        }
    }
}

extension ServerClient {
    func remotePlayers() async throws -> [WatchPlayerState] {
        let response = try await request("/api/v1/remote-players")
        let value = try response.body.object(allowing: ["players"])
        let players = try value.list("players", max: 64).map { raw -> WatchPlayerState in
            let value = try raw.object(allowing: ["id", "name", "title", "artist", "itemId", "state", "position", "duration", "audio"])
            let id = try value.text("id", max: 36, required: true)
            guard UUID(uuidString: id)?.uuidString.lowercased() == id else { throw ClientError.invalidResponse }
            let state = try value.text("state", max: 16, required: true)
            guard ["idle", "playing", "paused", "buffering"].contains(state) else { throw ClientError.invalidResponse }
            let player = WatchPlayerState(id: id, device: try value.text("name", max: 80, required: true),
                                          title: try value.text("title", max: 256), subtitle: try value.text("artist", max: 128),
                                          itemID: try value.text("itemId", max: 128), position: try value.number("position", max: 1_000_000_000),
                                          duration: try value.number("duration", max: 1_000_000_000),
                                          playing: state == "playing", audio: try value.flag("audio"))
            return try player.validated()
        }
        guard Set(players.map(\.id)).count == players.count else { throw ClientError.invalidResponse }
        return players
    }

    func updateRemotePlayer(id: String, state: WatchPlayerState) async throws -> RemotePlayerCommand? {
        guard UUID(uuidString: id)?.uuidString.lowercased() == id else { throw ClientError.invalidInput("The player ID is invalid.") }
        _ = try state.validated()
        let body: JSONValue = .object(["name": .string(state.device), "title": .string(state.title), "artist": .string(state.subtitle),
                                       "itemId": .string(state.itemID), "state": .string(state.active ? state.playing ? "playing" : "paused" : "idle"),
                                       "position": .number(state.position), "duration": .number(state.duration), "audio": .bool(state.audio)])
        let response = try await request("/api/v1/remote-players/\(id)", method: .put, body: body)
        let value = try response.body.object(allowing: ["command"])
        guard let command = value["command"] else { throw ClientError.invalidResponse }
        if case .null = command { return nil }
        return try RemotePlayerCommand(command)
    }

    func commandRemotePlayer(id: String, itemID: String, command: WatchRemoteRequest) async throws {
        guard UUID(uuidString: id)?.uuidString.lowercased() == id, command.target == id, let action = command.command else { throw ClientError.invalidInput("The player command is invalid.") }
        let validItemID = try Input.id(itemID)
        var body: [String: JSONValue] = ["command": .string(action), "itemId": .string(validItemID)]
        if let position = command.position { body["position"] = .number(position) }
        _ = try await request("/api/v1/remote-players/\(id)/commands", method: .post, body: .object(body), expected: [202])
    }
}

File: apps/player/apps/android/app/src/main/java/com/kinosail/player/core/RemotePlayersApi.kt
package com.kinosail.player.core

import com.kinosail.player.watchcore.WatchPlayer
import com.kinosail.player.watchcore.WatchRequest
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.put

internal class RemotePlayersApi(private val server: ServerAddress,
                                open: (java.net.URL) -> java.net.HttpURLConnection =
                                    { it.openConnection() as java.net.HttpURLConnection }) {
    private val api = ServerApi(server, open)

    fun list(token: String, viewerId: String): List<WatchPlayer> {
        val result = api.remotePlayers(token, viewerId).fields(setOf("players"), setOf("players"))
        val rows = result["players"] as? JsonArray ?: throw IllegalArgumentException(INVALID)
        require(rows.size <= 64) { INVALID }
        val players = rows.map { raw ->
            val value = raw.fields(PLAYER_FIELDS, PLAYER_FIELDS)
            WatchPlayer(value.text("id", 36), value.text("name", 80), value.text("title", 256),
                value.text("artist", 128), value.text("itemId", 128), value.text("state", 16),
                value.number("position"), value.number("duration"), value.flag("audio")).checked()
        }
        require(players.map(WatchPlayer::id).toSet().size == players.size && players.none { it.id == "phone" }) { INVALID }
        return players
    }

    fun update(player: WatchPlayer, token: String, viewerId: String): WatchRequest? {
        player.checked()
        require(player.id != "phone") { INVALID }
        val body = buildJsonObject {
            put("name", player.name); put("title", player.title); put("artist", player.subtitle)
            put("itemId", player.itemId); put("state", player.state)
            put("position", player.position); put("duration", player.duration); put("audio", player.audio)
        }
        val result = api.updateRemotePlayer(player.id, token, viewerId, body).fields(setOf("command"), setOf("command"))
        val command = result["command"]
        if (command == JsonNull) return null
        val value = (command ?: throw IllegalArgumentException(INVALID)).fields(COMMAND_FIELDS,
            setOf("command", "itemId"))
        val request = WatchRequest(player.id, value.text("itemId", 128), value.text("command", 16),
            value["position"]?.let { value.number("position") }).checked()
        require(player.active && request.itemId == player.itemId &&
            (request.position?.let { it <= player.duration } ?: true)) { INVALID }
        return request
    }

    fun command(player: WatchPlayer, request: WatchRequest, token: String, viewerId: String) {
        player.checked(); request.checked()
        val position = request.position
        require(player.id != "phone" && player.active && request.target == player.id &&
            request.itemId == player.itemId && (position == null || position <= player.duration)) { INVALID }
        val body = buildJsonObject {
            put("command", requireNotNull(request.command)); put("itemId", player.itemId)
            request.position?.let { put("position", it) }
        }
        api.commandRemotePlayer(player.id, token, viewerId, body)
    }

    private companion object {
        const val INVALID = "The Server returned an invalid remote player response."
        val PLAYER_FIELDS = setOf("id", "name", "title", "artist", "itemId", "state", "position", "duration", "audio")
        val COMMAND_FIELDS = setOf("command", "itemId", "position")
    }

    private fun JsonElement.fields(allowed: Set<String>, required: Set<String>): JsonObject {
        val value = this as? JsonObject
        require(value != null && value.keys.all(allowed::contains) && value.keys.containsAll(required)) { INVALID }
        return value
    }

    private fun JsonObject.text(key: String, max: Int): String {
        val value = this[key] as? JsonPrimitive
        require(value != null && value.isString && value.content.toByteArray(Charsets.UTF_8).size <= max &&
            value.content.none(Char::isISOControl)) { INVALID }
        return value.content
    }

    private fun JsonObject.number(key: String): Double {
        val value = (this[key] as? JsonPrimitive)?.takeIf { !it.isString }?.doubleOrNull
        require(value != null && value.isFinite()) { INVALID }
        return value
    }

    private fun JsonObject.flag(key: String): Boolean {
        val value = (this[key] as? JsonPrimitive)?.takeIf { !it.isString }?.booleanOrNull
        require(value != null) { INVALID }
        return value
    }
}
